subsys/nrf_security/src/drivers/nrf_cc3xx/Kconfig

```
@@ -1,262 +1,532 @@
 #
-# Copyright (c) 2021-2022 Nordic Semiconductor
+# Copyright (c) 2021 - 2023 Nordic Semiconductor
 #
 # SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
 #
 
-menuconfig PSA_CRYPTO_DRIVER_CC3XX
+# This invisible menu helps hiding these not user selectable options
+# from menuconfig even when show-all mode is enabled
+menu "PSA NEED CC3XX - invisible"
+visible if 0
+
+# CC3xx AEAD Driver
+
+config PSA_NEED_CC3XX_AES_CCM
 	bool
-	prompt "PSA CryptoCell Driver" if !PSA_PROMPTLESS
-	depends on CRYPTOCELL_USABLE
-	default n
-	select NRF_CC3XX_PLATFORM if !BUILD_WITH_TFM
-	help
-	  Enable PSA Driver for CryptoCell
+	default y
+	select PSA_ACCEL_AES128_CCM
+	select PSA_ACCEL_AES192_CCM if HAS_HW_NRF_CC312
+	select PSA_ACCEL_AES256_CCM if HAS_HW_NRF_CC312
+	depends on PSA_WANT_ALG_CCM
+	depends on PSA_WANT_KEY_TYPE_AES
+	depends on PSA_USE_CC3XX_AEAD_DRIVER
 
-if PSA_CRYPTO_DRIVER_CC3XX
+config PSA_NEED_CC3XX_AES_GCM
+	bool
+	default y
+	select PSA_ACCEL_AES128_GCM if HAS_HW_NRF_CC312
+	select PSA_ACCEL_AES192_GCM if HAS_HW_NRF_CC312
+	select PSA_ACCEL_AES256_GCM if HAS_HW_NRF_CC312
+	depends on PSA_WANT_ALG_GCM
+	depends on PSA_WANT_KEY_TYPE_AES
+	depends on PSA_USE_CC3XX_AEAD_DRIVER
 
-config PSA_CRYPTO_DRIVER_HAS_CIPHER_SUPPORT_CC3XX
+config PSA_NEED_CC3XX_CHACHA20_POLY1305
 	bool
 	default y
-	depends on PSA_CRYPTO_DRIVER_ALG_CBC_NO_PADDING_CC3XX || \
-		   PSA_CRYPTO_DRIVER_ALG_CBC_PKCS7_CC3XX || \
-		   PSA_CRYPTO_DRIVER_ALG_ECB_NO_PADDING_CC3XX || \
-		   PSA_CRYPTO_DRIVER_ALG_CTR_CC3XX || \
-		   PSA_CRYPTO_DRIVER_ALG_OFB_CC3XX
+	select PSA_ACCEL_CHACHA20_POLY1305
+	depends on PSA_WANT_ALG_CHACHA20_POLY1305
+	depends on PSA_USE_CC3XX_AEAD_DRIVER
 
-config PSA_CRYPTO_DRIVER_HAS_AEAD_SUPPORT_CC3XX
+config PSA_NEED_CC3XX_AEAD_DRIVER
 	bool
 	default y
-	depends on PSA_CRYPTO_DRIVER_ALG_CCM_CC3XX || \
-		   PSA_CRYPTO_DRIVER_ALG_GCM_CC3XX || \
-		   PSA_CRYPTO_DRIVER_ALG_CHACHA20_POLY1305_CC3XX
+	depends on PSA_NEED_CC3XX_AES_CCM || PSA_NEED_CC3XX_AES_GCM || PSA_NEED_CC3XX_CHACHA20_POLY1305
 
-config PSA_CRYPTO_DRIVER_HAS_MAC_SUPPORT_CC3XX
+# CC3xx Cipher Driver
+
+config PSA_NEED_CC3XX_AES_CTR
 	bool
 	default y
-	depends on PSA_CRYPTO_DRIVER_ALG_CMAC_CC3XX	|| \
-		   PSA_CRYPTO_DRIVER_ALG_HMAC_CC3XX
+	select PSA_ACCEL_AES128_CTR
+	select PSA_ACCEL_AES192_CTR if HAS_HW_NRF_CC312
+	select PSA_ACCEL_AES256_CTR if HAS_HW_NRF_CC312
+	depends on PSA_WANT_ALG_CTR
+	depends on PSA_WANT_KEY_TYPE_AES
```

```suggestion
	depends on PSA_WANT_KEY_TYPE_AES
	depends on PSA_WANT_AES_KEY_SIZE_128 if HAS_HW_NRF_CC310
```

For CC310, we should only use CC3XX for AES CTR when the requested key size is 128.

When CC3XX is requested, and AES_CTR is requested with a key size of 128, it looks like PSA_NEED_CC3XX_AES_CTR becomes set, but AFAICT from the Oberon docs it should be unset for this case.
AFAICT PSA_NEED_CC3XX_AES_CTR will be set when it should be unset for this case.

---

If we don't add this depends then for this case, the PSA_NEED_CC3XX_CIPHER_DRIVER option will become set and we will get a driver in our build that is just dead code.

Please correct me if I'm wrong.

---

If you wish we can add this dead code optimization to jira.

---

This optimization has merit. I see no problem adding it

---

I will add this for AES and RSA keys indeed.